azure/Guidelines.md

```diff
@@ -0,0 +1,865 @@
+>
+> :white_check_mark: **DO** support the Repeatability-Request-ID & Repeatability-First-Sent request headers if the action needs to be idempotent if retries occur.
+
+> :ballot_box_with_check: **YOU SHOULD** use a verb to name your action.
+
+
+### Collections
+> :white_check_mark: **DO** structure the response to a list operation as an object with a top-level array field containing the set (or subset) of resources.
+
+> :ballot_box_with_check: **YOU SHOULD** support paging today if there is ever a chance in the future that the number of items can grow to be very large. 
+> NOTE: It is a breaking change to add paging in the future
+
+> :heavy_check_mark: **YOU MAY** expose an operation that lists your resources by supporting a GET method with a URL to a resource-collection (as opposed to a resource-id).
+
+**Example Response Body** 
+```
+{
+    "value": [
+       { "id": "Item 01", "etag": "0xabc", "price": 99.95, "sizes": null },
+       { … },
+       { … },
+       { "id": "Item 99", "etag": "0xdef", "price": 59.99, "sizes": null }
+    ],
+    "nextLink": "{opaqueUrl}"
+ }
+```
+
+> :white_check_mark: **DO** include the id field and etag field (if supported) for each item as this allows the customer to modify the item in a future operation.
+> 
+> :white_check_mark: **DO** clearly document that resources may be skipped or duplicated across pages of a paginated collection unless the operation has made special provisions to prevent this (like taking a time-expiring snapshot of the collection).
+>
+> :white_check_mark: **DO** return a ```nextLink``` field with an absolute URL that the client can GET in order to retrieve the next page of the collection.
+
+> :ballot_box_with_check: **YOU SHOULD** use ```value``` as the name of the top-level array field unless a more appropriate name is available.
+
+> :no_entry: **DO NOT** return the ```nextLink``` field at all when returning the last page of the collection. 
+>
+> :no_entry: **DO NOT** ever return a ```nextLink``` field with a value of null.
+
+#### Query options
+> :heavy_check_mark: **YOU MAY** support the following query parameters allowing customers to control the list operation:
+
+Parameter&nbsp;name | Type | Description
+------------------- | ---- | -----------
+_filter_       | string            | an expression on the resource type that selects the resources to be returned
+_orderby_      | string&nbsp;array | a list of expressions that specify the order of the returned resources
+_skip_         | integer           | an offset into the collection of the first resource to be returned
+_top_          | integer           | the maximum number of resources to return from the collection
+_maxpagesize_  | integer           | the maximum number of resources to include in a single response
+_select_       | string&nbsp;array | a list of field names to be returned for each resource
+_expand_       | string&nbsp;array | a list of the related resources to be included in line with each resource
+
```

Comment:
NOTE: The orderby parameter can be extremely expensive for a service to implement as it must retrieve all items to sort them. And if the operation supports paging (which is likely), then a client request to get another page may have to retrieve all items and sort them again to determine which items are on the desired page. Only support orderby if customer scenarios really demand it and the service is confident that it can support it in perpetuity (even if the backing storage service changes someday).